### PR TITLE
Overview/state-scripts

### PR DESCRIPTION
**Pull request by oleorhagen**
New Overview/State-scripts section

### CODE REVIEW

File: 02.Overview/08.State-scripts/docs.md [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+
+In the case that a state script returns an error code during execution, the
+update fails, and the _ArtifactFailure_ scripts are run, before an error is
+reported to the Mender server.
+
+## Logging
+
+State scripts can have a log of their execution collected and pushed to the
+Mender server in case of an error. This is done through writing to _stderr_
+during script execution.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:35]: > State scripts can have a log of their execution collected and pushed to the Mender server in case of an error.

Would reword this slightly,

The Mender client collects logs during an update which it will send to the Mender server in case of an failure. The output from state scripts are included in these logs. This can be useful when troubleshooting intermittent update failures.


> This is done through writing to _stderr_ during script execution.

Implementation detail. Covered in https://docs.mender.io/2.3/artifacts/state-scripts#example-use-cases

File: 02.Overview/08.State-scripts/docs.md [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+* **ArtifactCommit**: If the new image in the passive partition passes all the integrity checks, the Mender client will mark the update as successful and continue running from this partition. The commit makes the update persistent. A pre-script here can add custom integrity checks beyond what the Mender client provides out of the box.
+* **ArtifactRollback**: If the new update fails any of the above mentioned integrity checks, the Mender client will revert the update payload. At this point, scripts in this state can help revert migrations done as a part of the update.
+* **ArtifactRollbackReboot**: Some update types - most notably rootfs images - require a reboot, if any of the integrity checks fail. This state runs right before the client reboots back (i.e., rollback) into the old image partition.
+* **ArtifactFailure**: if any of the States with an "Artifact" prefix fail, execute this state. Note: this state always runs after the ArtifactRollback and the ArtifactRollbackReboot states.
+
+Most of the states also have an _Error_ transition which runs if some error
+occurs while executing any script for the given state.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:50]: Not sure about this. If we are to include it, it should be together with the diagram.

So we have a couple of options:
1. Remove this, then we should also remove the diagram
2. Include, but move closer to the diagram (keep duplicate of what is in https://docs.mender.io/2.3/artifacts/state-scripts#example-use-cases)
3. Include, move closer to the diagram and remove this info from https://docs.mender.io/2.3/artifacts/state-scripts#example-use-cases

I am leaning towards 2. @estenberg ?

File: 02.Overview/08.State-scripts/docs.md [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+taxonomy:
+    category: docs
+    label: reference
+---
+
+State scripts offers customization of the Mender state-machine during an update.
+The Mender Client can run scripts before and after every state in the Mender
+state-machine. This means that through the use of state scripts you can
+customize how Mender behaves during an update. Some common use-cases include:
+turning on and off the network connectivity before and after checking for an
```
[the remark this answers is not in the document]
Comment (estenberg) [reply 02.Overview/08.State-scripts/docs.md:12]: Agree, I think it's important to lead with the use cases as "teasers" here.

File: 02.Overview/08.State-scripts/docs.md (hunk 2 above)
Comment (estenberg) [reply 02.Overview/08.State-scripts/docs.md:50]: I think 2) is fine as well. The idea is to inspire what is possible, and the diagram with a short description of each state helps with this I think.

File: 02.Overview/08.State-scripts/docs.md [hunk 4]
```diff
@@ -0,0 +1,68 @@
...
+  actions of the Mender state machine, during an update installation. Artifact
+  scripts are easy to distinguish from root file system scripts, as they are
+  prefixed with _Artifact_.
+  
+**Root file system** and **Artifact** scripts serve different use-cases. Since
+Mender allows some scripts to run prior to the download of the Artifact and as
+such, these are not run from the Artifact which contains them. This means that
+scripts with the names `Idle`, `Sync` or `Download`, get installed on the root
+file system permanently after an update. Scripts starting with the name
+`Artifact`, are only run during an update process, and are since deleted.
```
[the remark this answers is not in the document]
Comment (estenberg) [reply 02.Overview/08.State-scripts/docs.md:68]: Agree.

File: 02.Overview/08.State-scripts/docs.md [hunk 5]
```diff
@@ -0,0 +1,68 @@
...
+State scripts are run as pre- and post-scripts to every state in the Mender
+state machine shown above. The state in which a given script is run is
+determined by the script name and index, as detailed in the [State
+script](../../artifacts/state-scripts) section.
+
+## Errors
+
+In the case that a state script returns an error code during execution, the
+update fails, and the _ArtifactFailure_ scripts are run, before an error is
+reported to the Mender server.
```
[the remark this answers is not in the document]
Comment (estenberg) [reply 02.Overview/08.State-scripts/docs.md:29]: Makes sense, but to me error handling is more for when you are implementing (not this section).

File: 02.Overview/08.State-scripts/docs.md (hunk 1 above)
Comment (estenberg) [reply 02.Overview/08.State-scripts/docs.md:35]: Also here, I think I would care about logging only when implementing this (not here).

File: 02.Overview/08.State-scripts/docs.md (hunk 5 above)
[the remark this answers is not in the document]
Comment (mirzak) [reply 02.Overview/08.State-scripts/docs.md:29]: Yeah, but this is not about error handling and is more what options you have at a high level, 
This is a compliment to `The Mender Client has the ability to run pre- and postinstall scripts`, and should highlight **what** you can do in the pre/post install step, it should not cover **how**.

E.g if I know that I can trigger an update failure in my state-script, this opens up for certain use-cases. I think it is worth to mention here.

File: 02.Overview/08.State-scripts/docs.md (hunk 1 above)
Comment (mirzak) [reply 02.Overview/08.State-scripts/docs.md:35]: I think there some value in highlighting the troubleshooting possibilities, but do not feel strongly about this.

**Comment by oleorhagen** [comment]
@mirzak Ready for a second round of slaughter

**Comment by merlin-northern** [comment]
havent we agree that Drew reviews all?

File: 02.Overview/08.State-scripts/docs.md [hunk 6]
```diff
@@ -0,0 +1,57 @@
...
+
+## Errors
+
+Errors returned from a state script controls the execution of the Mender state
+machine. Depending on the return code from the script the Mender client will:
+
+* Move on to the next state, if successful.
+* Retry this state at a later time.
+* Mark the update as failed, and report the error to the server.
+
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:52]: Would not label this `Errors`. Maybe merge this with the `Execution` section and remove

```
Errors returned from a state script controls the execution of the Mender state machine.
```

Would also add a link to "learn more"

File: 02.Overview/08.State-scripts/docs.md [hunk 7]
```diff
@@ -0,0 +1,57 @@
...
+## Logging
+
+The Mender client collects logs during an update. In case of a failure, these
+logs are sendt to the Mender server. The output from state scripts are included
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:56]: `sendt` -> `sent`

File: 02.Overview/08.State-scripts/docs.md [hunk 8]
```diff
@@ -0,0 +1,57 @@
...
+in these logs. This is useful when troubleshooting intermittent update failures.
```
Comment (mirzak) [inline 02.Overview/08.State-scripts/docs.md:57]: `intermittent update failures` -> `update failures`.

I know I suggested this text earlier, but it feels to complex when looking at it again. Update if you agree.

File: 02.Overview/08.State-scripts/docs.md (hunk 7 above)
Comment (oleorhagen) [reply 02.Overview/08.State-scripts/docs.md:56]: haha, a wee bit of norwegian :see_no_evil:

File: 02.Overview/08.State-scripts/docs.md [hunk 9]
```diff
@@ -53,5 +50,10 @@ machine. Depending on the return code from the script the Mender client will:
 ## Logging
 
 The Mender client collects logs during an update. In case of a failure, these
-logs are sendt to the Mender server. The output from state scripts are included
+logs are sent to the Mender server. The output from state scripts are included
 in these logs. This is useful when troubleshooting intermittent update failures.
+
+## Furter reading
+
+To learn more, have a look at the [State script
```
Comment (estenberg) [inline 02.Overview/08.State-scripts/docs.md:58]: would say [tutorial on how to implement State scripts](...) to make it more clear how that doc is different than the one I'm currently reading.

File: 02.Overview/08.State-scripts/docs.md [hunk 10]
```diff
@@ -0,0 +1,59 @@
...
+
+The Mender Client has the ability to run pre- and postinstall scripts, before
+and after it writes the root file system. However, Mender state scripts are more
+general and useful than pre/postinstall scripts because they run between
+any state transition, not just (before/after) the install state.
+
+Common use cases are:
+
+* Enable/disable network connectivity before and after checking for an update to reduce bandwidth.
+* Migrate persistent data during a software upgrade.
```
Comment (estenberg) [inline 02.Overview/08.State-scripts/docs.md:16]: "software update" is what we commonly use (not upgrade)
